ProcessRGB.cpp

```diff
@@ -3266,16 +3266,20 @@ etcpak_force_inline static int16x8_t WidenMultiplier_EAC_NEON( int16x8_t multipl
 
 #endif
 
-static etcpak_force_inline uint64_t ProcessAlpha_ETC2( const uint8_t* src )
+static etcpak_force_inline uint64_t ProcessAlpha_ETC2( const uint8_t* src, bool checkSolid = true )
```

Comment:
What is the performance impact of adding these checks?

Reply:
On a 16K RG texture:

checkSolid = false:
Median compression time for 9 runs: 1490.763 ms (180.066 Mpx/s) single threaded
Median compression time for 9 runs: 125.559 ms (2137.923 Mpx/s) multi threaded (24 cores)

checkSolid = true:
Median compression time for 9 runs: 1211.941 ms (221.492 Mpx/s) single threaded
Median compression time for 9 runs: 103.266 ms (2599.456 Mpx/s) multi threaded (24 cores)

So unfortunately, the performance cost is not negligible. This was done since, to my knowledge, in the 11-bit formats multiplier 0 is interpreted as 1, so during decoding a solid value of 255 would get expanded to 2044 (base * 8 + 4) and -3 * 1 would be added to it, resulting in 2041. This meant that the decoded value would never reach 1.0.

There may be a more performance-friendly workaround available though, so I will look into it.

Reply:
I am asking in the context of existing code. You're adding a bunch of branches, so I'm more interested in how the behavior changes with and without the `checkSolid` branches. So my question here is: are the compilers (gcc / clang / MSVC) able to generate two instances of this function, one with `checkSolid` always true and the second always false, effectively omitting the branches? I know that in some cases the optimizer is able to do this, but is it also true here? Viewing the disassembly would be a sure way to check this.

If not, it might be worth making this function a template with `checkSolid` as a template parameter, which would force the output of two different function instances.

Reply:
After checking the assembly, it looks like compilers are indeed able to generate different instances of the function (with optimizations enabled). Adding a template to the function might still be worth it, just to ensure it always works correctly.